File: src/pixel_art_mcp/authoring.py
"""Public, typed pixel authoring contract. The engine runner uses the stdlib-only helper."""

from typing import Annotated, Any, Literal
from uuid import UUID

from pydantic import Field, StrictBool, StrictInt, model_validator

from pixel_art_mcp.drawing import PixelDrawing, PixelModel, PixelRow, Symbol
from pixel_art_mcp.models import AssetSpec, DomainError, Model
from pixel_art_mcp.pixel_art import Canvas, PixelArt

AUTHORING_VERSION = 1
Color = Annotated[str, Field(pattern=r"^#[0-9a-fA-F]{6}$")]


class PixelPose(PixelModel):
    """One view/frame patch. Dots reveal lower layers; rows are not resampled."""

    angle: Literal[0, 90, 180, 270] = Field(
        description="0 front/down, 90 right, 180 back/up, 270 left. Use only configured views."
    )
    rows: list[PixelRow] | None = Field(
        default=None,
        min_length=1,
        max_length=512,
        description="Equal-width pixel rows, top to bottom. Letters/digits refer to palette "
        "symbols; '.' is transparent. Supply exactly one of rows or drawing. Prefer drawing "
        "for large shapes; literal rows are suited to small motifs.",
    )
    drawing: PixelDrawing | None = Field(
        default=None,
        description="Numeric rect/line/stamp commands using mandatory Canvas helpers. Alternative "
        "to rows, not additional pixels. Source reads return canonical raster rows.",
    )
    frame: StrictInt | None = Field(
        default=None,
        ge=0,
        le=100_000,
        description="Source frame, not playback index. null is this view's default; an exact "
        "frame replaces the entire default patch. Without a default, unspecified "
        "frames hide the layer.",
    )
    x: StrictInt = Field(
        default=0,
        ge=-512,
        le=512,
        description="Native pixel column of the patch's top-left.",
    )
    y: StrictInt = Field(
        default=0,
        ge=-512,
        le=512,
        description="Native pixel row of the patch's top-left; positive goes down.",
    )
    min_pixels: StrictInt = Field(
        default=0,
        ge=0,
        le=262_144,
        description="Advisory minimum pixels still visible after compositing all layers. Use "
        "for identifying details. A failed budget is an inspection finding, not a "
        "structural error.",
    )
    connected: StrictBool = Field(
        default=False,
        description="Require one four-connected visible cluster as an advisory feature check. "
        "False permits separated shapes such as a pair of eyes.",
    )

    @model_validator(mode="after")
    def rectangular(self) -> "PixelPose":
        if (self.rows is None) == (self.drawing is None):
            raise ValueError("Supply exactly one of rows or drawing")
        if self.rows is not None:
            Canvas.from_rows(self.rows)
        return self

    def canvas(self) -> Canvas:
        if self.drawing is not None:
            return self.drawing.canvas()
        assert self.rows is not None
        return Canvas.from_rows(self.rows)


class PixelLayer(PixelModel):
    """Stable name plus complete per-view poses. Document order is back to front."""

    name: str = Field(
        min_length=1,
        max_length=100,
        description="Unique layer name, e.g. barrel, faucet, gauge, rain. Retain names when "
        "editing.",
    )
    poses: list[PixelPose] = Field(
        min_length=1,
        max_length=256,
        description="At most one pose per (angle, frame), including a possible null-frame "
        "default for each view.",
    )

    @model_validator(mode="after")
    def unique_poses(self) -> "PixelLayer":
        if len({(p.angle, p.frame) for p in self.poses}) != len(self.poses):
            raise ValueError("Duplicate (angle, frame) poses in layer " + self.name)
        return self


class PixelDefinition(PixelModel):
    """Complete editable source; write replaces it atomically, never merges it."""

    version: Literal[1] = 1
    palette: dict[Symbol, Color] = Field(
        min_length=2,
        max_length=64,
        description="Distinct symbol-to-#RRGGBB colors, e.g. {D: #293039, G: #f3cf65}. Must "
        "fit configure_asset.colors and match configure_asset.palette if supplied. "
        "This is the whole job's authoritative palette.",
    )
    layers: list[PixelLayer] = Field(
        min_length=1,
        max_length=128,
        description="Complete ordered layer list, painted back to front. Omitted old layers "
        "are deleted. Separate a static body from small animated patches; "
        "views/dimensions come from configure_asset, not this document.",
    )

    @model_validator(mode="after")
    def valid_document(self) -> "PixelDefinition":
        if len({layer.name for layer in self.layers}) != len(self.layers):
            raise ValueError("Layer names must be unique")
        if len({color.lower() for color in self.palette.values()}) != len(self.palette):
            raise ValueError("Palette colors must be distinct")
        pixels = 0
        paint_operations = 0
        for layer in self.layers:
            for pose in layer.poses:
                if pose.drawing is not None:
                    pixels += pose.drawing.width * pose.drawing.height
                    paint_operations += pose.drawing.cost()
                    symbols = pose.drawing.symbols()
                else:
                    assert pose.rows is not None
                    pixels += sum(map(len, pose.rows))
                    symbols = set("".join(pose.rows)) - {"."}
                if symbols - self.palette.keys():
                    raise ValueError(f"Unknown palette symbol in layer {layer.name!r}")
        if pixels > 262_144:
            raise ValueError("Definition exceeds 262144 authored pixel cells; reuse default poses")
        if paint_operations > 1_048_576:
            raise ValueError("Definition exceeds 1048576 drawing paint operations")
        return self

    def to_art(self, layouts: list[dict[str, Any]]) -> PixelArt:
        art = PixelArt(self.palette, {v["angle"]: (v["width"], v["height"]) for v in layouts})
        for layer in self.layers:
            for pose in layer.poses:
                art.layer(
                    layer.name,
                    canvas=pose.canvas(),
                    **pose.model_dump(exclude={"rows", "drawing"}),
                )
        return art

    @classmethod
    def from_art(cls, data: dict[str, Any]) -> "PixelDefinition":
        return cls.model_validate({k: v for k, v in data.items() if k != "views"})


class PixelArtSource(Model):
    """Resumable source. Send only definition back to write_pixel_art."""

    contract_version: Literal[1] = 1
    project_id: UUID
    revision_id: UUID
    definition: PixelDefinition
    authored_views: dict[str, list[int]] = Field(
        description="Canvas dimensions saved with this revision. Reconfiguration may require "
        "adapting the definition before rendering."
    )
    configuration_id: UUID | None


class PoseTarget(PixelModel):
    layer: str = Field(min_length=1, max_length=100, description="Existing, exact layer name.")
    angle: Literal[0, 90, 180, 270]
    frame: StrictInt | None = Field(
        ge=0,
        le=100_000,
        description="Exact stored pose key. null selects the view default, not all frames.",
    )


class MovePose(PoseTarget):
    """Move one existing patch without resending rows or changing any other property."""

    op: Literal["move_pose"]
    x: StrictInt = Field(ge=-512, le=512, description="New absolute native x, not a delta.")
    y: StrictInt = Field(ge=-512, le=512, description="New absolute native y, not a delta.")


class SetPose(PixelModel):
    """Replace an exact pose in place, or append it; the named layer must already exist."""

    op: Literal["set_pose"]
    layer: str = Field(min_length=1, max_length=100)
    pose: PixelPose = Field(
        description="Complete patch. Omitted properties use PixelPose defaults."
    )


class DeletePose(PoseTarget):
    """Delete an exact stored pose. Deleting an override may reveal its view default."""

    op: Literal["delete_pose"]


class SetLayer(PixelModel):
    """Replace a named layer at its current order, or append a new topmost layer."""

    op: Literal["set_layer"]
    layer: PixelLayer = Field(description="Complete layer; omitted previous poses are deleted.")


class DeleteLayer(PixelModel):
    """Delete an existing named layer. Missing targets fail the entire edit batch."""

    op: Literal["delete_layer"]
    name: str = Field(min_length=1, max_length=100)


class SetPalette(PixelModel):
    """Replace the palette; all retained poses must use symbols from the final palette."""

    op: Literal["set_palette"]
    palette: dict[Symbol, Color] = Field(min_length=2, max_length=64)


PixelEdit = Annotated[
    MovePose | SetPose | DeletePose | SetLayer | DeleteLayer | SetPalette,
    Field(discriminator="op"),
]
PixelEdits = Annotated[
    list[PixelEdit],
    Field(
        min_length=1,
        max_length=128,
        description="Ordered atomic operations; only the final document is validated.",
    ),
]


def apply_pixel_edits(definition: PixelDefinition, edits: PixelEdits) -> PixelDefinition:
    """Work on a detached document so any failure leaves the original source untouched."""
    data = definition.model_dump()
    layers = data["layers"]
    for index, edit in enumerate(edits):
        if isinstance(edit, SetPalette):
            data["palette"] = dict(edit.palette)
            continue
        if isinstance(edit, SetLayer):
            name = edit.layer.name
        elif isinstance(edit, DeleteLayer):
            name = edit.name
        else:
            name = edit.layer
        layer_index = next((i for i, layer in enumerate(layers) if layer["name"] == name), None)
        if isinstance(edit, SetLayer):
            if layer_index is None:
                layers.append(edit.layer.model_dump())
            else:
                layers[layer_index] = edit.layer.model_dump()
            continue
        if layer_index is None:
            raise ValueError(f"edits[{index}] {edit.op}: unknown layer {name!r}")
        if isinstance(edit, DeleteLayer):
            layers.pop(layer_index)
            continue
        poses = layers[layer_index]["poses"]
        target = edit.pose if isinstance(edit, SetPose) else edit
        pose_index = next(
            (
                i
                for i, pose in enumerate(poses)
                if (pose["angle"], pose["frame"]) == (target.angle, target.frame)
            ),
            None,
        )
        if isinstance(edit, SetPose):
            if pose_index is None:
                poses.append(edit.pose.model_dump())
            else:
                poses[pose_index] = edit.pose.model_dump()
        elif pose_index is None:
            raise ValueError(
                f"edits[{index}] {edit.op}: no stored pose ({target.angle}, {target.frame}) "
                f"in {name!r}; defaults are not implicit edit targets"
            )
        elif isinstance(edit, DeletePose):
            poses.pop(pose_index)
        else:
            poses[pose_index].update(x=edit.x, y=edit.y)
    return PixelDefinition.model_validate(data)


def validated_art(data: dict[str, Any], options: dict[str, Any]) -> PixelArt:
    try:
        PixelDefinition.from_art(data)
        art = PixelArt.from_dict(data)
        spec = AssetSpec.model_validate(options["asset"])
        art.validate_target(options["asset_layouts"], options["frame_sequence"], spec.model_dump())
        return art
    except (ValueError, KeyError, TypeError) as exc:
        raise DomainError(f"Invalid pixel-art definition: {exc}") from exc

File: packages/schema/src/authoring.ts
/**
 * Port of `src/pixel_art_mcp/authoring.py`: the versioned pixel authoring contract. The
 * `PixelPose.canvas()`/`PixelDefinition.to_art()`/`from_art()` methods that build a real
 * `PixelArt` (from `pixel_art.py`) are Phase-3 (`packages/pixel-core`) territory and are not
 * ported here -- see the note on `validatedArt` below for the one place that boundary shows up.
 */

import { z } from "zod";
import {
  arrayField,
  DomainError,
  intField,
  literalField,
  modelObject,
  recordField,
  stringField,
  taggedUnion,
  uuidField,
  valueError,
} from "./errors.js";
import {
  PixelDrawingSchema,
  pixelDrawingCost,
  pixelDrawingSymbols,
  rectangularRowsError,
  SYMBOL_PATTERN,
  PixelRow,
} from "./drawing.js";
import { AssetSpecSchema } from "./models.js";

export const AUTHORING_VERSION = 1;

/** `Color = Annotated[str, Field(pattern=r"^#[0-9a-fA-F]{6}$")]` */
export const COLOR_PATTERN = /^#[0-9a-fA-F]{6}$/;
export const ColorSchema = stringField({
  pattern: COLOR_PATTERN,
  patternText: "^#[0-9a-fA-F]{6}$",
});

const ANGLE_VALUES = [0, 90, 180, 270] as const;
export type Angle = (typeof ANGLE_VALUES)[number];
export const AngleSchema = literalField(ANGLE_VALUES);

/** Formats a nullable int the way an f-string interpolates it in Python: `None`, not `null`. */
function pyNone(value: number | null): string {
  return value === null ? "None" : String(value);
}

/** Formats a Python `!r` string repr closely enough for these error messages' purposes. */
function pyRepr(value: string): string {
  if (value.includes("'") && !value.includes('"')) return `"${value}"`;
  return `'${value.replace(/'/g, "\\'")}'`;
}

/** `PixelPose(PixelModel)` */
export const PixelPoseSchema = modelObject(
  {
    angle: AngleSchema.describe(
      "0 front/down, 90 right, 180 back/up, 270 left. Use only configured views.",
    ),
    rows: arrayField(PixelRow, {
      minLength: 1,
      maxLength: 512,
      description:
        "Equal-width pixel rows, top to bottom. Letters/digits refer to palette " +
        "symbols; '.' is transparent. Supply exactly one of rows or drawing. Prefer drawing " +
        "for large shapes; literal rows are suited to small motifs.",
    })
      .nullable()
      .default(null),
    drawing: PixelDrawingSchema.describe(
      "Numeric rect/line/stamp commands using mandatory Canvas helpers. Alternative " +
        "to rows, not additional pixels. Source reads return canonical raster rows.",
    )
      .nullable()
      .default(null),
    frame: intField({
      ge: 0,
      le: 100_000,
      description:
        "Source frame, not playback index. null is this view's default; an exact " +
        "frame replaces the entire default patch. Without a default, unspecified " +
        "frames hide the layer.",
    })
      .nullable()
      .default(null),
    x: intField({
      ge: -512,
      le: 512,
      description: "Native pixel column of the patch's top-left.",
    }).default(0),
    y: intField({
      ge: -512,
      le: 512,
      description: "Native pixel row of the patch's top-left; positive goes down.",
    }).default(0),
    min_pixels: intField({
      ge: 0,
      le: 262_144,
      description:
        "Advisory minimum pixels still visible after compositing all layers. Use " +
        "for identifying details. A failed budget is an inspection finding, not a " +
        "structural error.",
    }).default(0),
    connected: z
      .boolean()
      .default(false)
      .describe(
        "Require one four-connected visible cluster as an advisory feature check. " +
          "False permits separated shapes such as a pair of eyes.",
      ),
  },
  (value, ctx) => {
    if ((value.rows === null) === (value.drawing === null)) {
      valueError(ctx, "Supply exactly one of rows or drawing");
      return;
    }
    if (value.rows !== null) {
      const error = rectangularRowsError(value.rows);
      if (error) valueError(ctx, error);
    }
  },
);
export type PixelPose = z.infer<typeof PixelPoseSchema>;

/** `PixelLayer(PixelModel)` */
export const PixelLayerSchema = modelObject(
  {
    name: stringField({
      minLength: 1,
      maxLength: 100,
      description:
        "Unique layer name, e.g. barrel, faucet, gauge, rain. Retain names when editing.",
    }),
    poses: arrayField(PixelPoseSchema, {
      minLength: 1,
      maxLength: 256,
      description:
        "At most one pose per (angle, frame), including a possible null-frame " +
        "default for each view.",
    }),
  },
  (value, ctx) => {
    const seen = new Set<string>();
    for (const pose of value.poses) {
      const key = `${String(pose.angle)}:${String(pose.frame)}`;
      seen.add(key);
    }
    if (seen.size !== value.poses.length) {
      valueError(ctx, `Duplicate (angle, frame) poses in layer ${value.name}`);
    }
  },
);
export type PixelLayer = z.infer<typeof PixelLayerSchema>;

/** `PixelDefinition(PixelModel)` */
export const PixelDefinitionSchema = modelObject(
  {
    version: literalField([1] as const).default(1),
    palette: recordField(ColorSchema, {
      minLength: 2,
      maxLength: 64,
      keyPattern: SYMBOL_PATTERN,
      keyPatternText: "^[A-Za-z0-9]$",
      description:
        "Distinct symbol-to-#RRGGBB colors, e.g. {D: #293039, G: #f3cf65}. Must " +
        "fit configure_asset.colors and match configure_asset.palette if supplied. " +
        "This is the whole job's authoritative palette.",
    }),
    layers: arrayField(PixelLayerSchema, {
      minLength: 1,
      maxLength: 128,
      description:
        "Complete ordered layer list, painted back to front. Omitted old layers " +
        "are deleted. Separate a static body from small animated patches; " +
        "views/dimensions come from configure_asset, not this document.",
    }),
  },
  (value, ctx) => {
    const layerNames = new Set(value.layers.map((l) => l.name));
    if (layerNames.size !== value.layers.length) {
      valueError(ctx, "Layer names must be unique");
      return;
    }
    const paletteColors = new Set(Object.values(value.palette).map((c) => c.toLowerCase()));
    if (paletteColors.size !== Object.keys(value.palette).length) {
      valueError(ctx, "Palette colors must be distinct");
      return;
    }
    let pixels = 0;
    let paintOperations = 0;
    for (const layer of value.layers) {
      for (const pose of layer.poses) {
        let symbols: Set<string>;
        if (pose.drawing !== null) {
          const drawing = pose.drawing;
          pixels += drawing.width * drawing.height;
          paintOperations += pixelDrawingCost(drawing);
          symbols = pixelDrawingSymbols(drawing);
        } else {
          const rows = pose.rows;
          if (rows === null) {
            // PixelPoseSchema's own refine already enforces exactly one of rows/drawing;
            // unreachable for a pose that parsed successfully.
            throw new Error("Invariant violated: validated pose has neither rows nor drawing");
          }
          pixels += rows.reduce((sum, row) => sum + row.length, 0);
          symbols = new Set(
            rows
              .join("")
              .split("")
              .filter((c) => c !== "."),
          );
        }
        const unknown = [...symbols].filter((s) => !(s in value.palette));
        if (unknown.length > 0) {
          valueError(ctx, `Unknown palette symbol in layer ${pyRepr(layer.name)}`);
          return;
        }
      }
    }
    if (pixels > 262_144) {
      valueError(ctx, "Definition exceeds 262144 authored pixel cells; reuse default poses");
      return;
    }
    if (paintOperations > 1_048_576) {
      valueError(ctx, "Definition exceeds 1048576 drawing paint operations");
    }
  },
);
export type PixelDefinition = z.infer<typeof PixelDefinitionSchema>;

/** `PixelArtSource(Model)` */
export const PixelArtSourceSchema = modelObject({
  schema_version: literalField([1] as const).default(1),
  contract_version: literalField([1] as const).default(1),
  project_id: uuidField(),
  revision_id: uuidField(),
  definition: PixelDefinitionSchema,
  authored_views: recordField(arrayField(intField()), {
    description:
      "Canvas dimensions saved with this revision. Reconfiguration may require " +
      "adapting the definition before rendering.",
  }),
  configuration_id: uuidField().nullable(),
});
export type PixelArtSource = z.infer<typeof PixelArtSourceSchema>;

/** `PoseTarget(PixelModel)` -- inlined into each edit variant below (Zod has no base-class reuse
 * for discriminated union members beyond spreading the same field definitions). */
const poseTargetShape = {
  layer: stringField({ minLength: 1, maxLength: 100, description: "Existing, exact layer name." }),
  angle: AngleSchema,
  frame: intField({
    ge: 0,
    le: 100_000,
    description: "Exact stored pose key. null selects the view default, not all frames.",
  }).nullable(),
};

export const MovePoseSchema = modelObject({
  ...poseTargetShape,
  op: literalField(["move_pose"] as const),
  x: intField({ ge: -512, le: 512, description: "New absolute native x, not a delta." }),
  y: intField({ ge: -512, le: 512, description: "New absolute native y, not a delta." }),
});
export type MovePose = z.infer<typeof MovePoseSchema>;

export const SetPoseSchema = modelObject({
  op: literalField(["set_pose"] as const),
  layer: stringField({ minLength: 1, maxLength: 100 }),
  pose: PixelPoseSchema.describe("Complete patch. Omitted properties use PixelPose defaults."),
});
export type SetPose = z.infer<typeof SetPoseSchema>;

export const DeletePoseSchema = modelObject({
  ...poseTargetShape,
  op: literalField(["delete_pose"] as const),
});
export type DeletePose = z.infer<typeof DeletePoseSchema>;

export const SetLayerSchema = modelObject({
  op: literalField(["set_layer"] as const),
  layer: PixelLayerSchema.describe("Complete layer; omitted previous poses are deleted."),
});
export type SetLayer = z.infer<typeof SetLayerSchema>;

export const DeleteLayerSchema = modelObject({
  op: literalField(["delete_layer"] as const),
  name: stringField({ minLength: 1, maxLength: 100 }),
});
export type DeleteLayer = z.infer<typeof DeleteLayerSchema>;

export const SetPaletteSchema = modelObject({
  op: literalField(["set_palette"] as const),
  palette: recordField(ColorSchema, {
    minLength: 2,
    maxLength: 64,
    keyPattern: SYMBOL_PATTERN,
    keyPatternText: "^[A-Za-z0-9]$",
  }),
});
export type SetPalette = z.infer<typeof SetPaletteSchema>;

export type PixelEdit = MovePose | SetPose | DeletePose | SetLayer | DeleteLayer | SetPalette;

export const PixelEditSchema = taggedUnion("op", {
  move_pose: MovePoseSchema,
  set_pose: SetPoseSchema,
  delete_pose: DeletePoseSchema,
  set_layer: SetLayerSchema,
  delete_layer: DeleteLayerSchema,
  set_palette: SetPaletteSchema,
});

/** `PixelEdits = Annotated[list[PixelEdit], Field(min_length=1, max_length=128, ...)]` */
export const PixelEditsSchema = arrayField(PixelEditSchema, {
  minLength: 1,
  maxLength: 128,
  description: "Ordered atomic operations; only the final document is validated.",
});
export type PixelEdits = z.infer<typeof PixelEditsSchema>;

/**
 * Port of `apply_pixel_edits`: works on a deep-cloned plain object so a failure partway through
 * leaves the caller's `definition` untouched, then re-validates the whole result through
 * `PixelDefinitionSchema` exactly like the Python version re-validates through
 * `PixelDefinition.model_validate`.
 */
export function applyPixelEdits(definition: PixelDefinition, edits: PixelEdits): PixelDefinition {
  const data = structuredClone(definition) as {
    palette: Record<string, string>;
    layers: {
      name: string;
      poses: { angle: number; frame: number | null; x: number; y: number }[];
    }[];
  };
  const layers = data.layers;

  edits.forEach((edit, index) => {
    if (edit.op === "set_palette") {
      data.palette = { ...edit.palette };
      return;
    }
    const name =
      edit.op === "set_layer"
        ? edit.layer.name
        : edit.op === "delete_layer"
          ? edit.name
          : edit.layer;
    const layerIndex = layers.findIndex((layer) => layer.name === name);

    if (edit.op === "set_layer") {
      const clonedLayer = structuredClone(edit.layer);
      if (layerIndex === -1) layers.push(clonedLayer);
      else layers[layerIndex] = clonedLayer;
      return;
    }
    if (layerIndex === -1) {
      throw new Error(`edits[${index}] ${edit.op}: unknown layer ${pyRepr(name)}`);
    }
    if (edit.op === "delete_layer") {
      layers.splice(layerIndex, 1);
      return;
    }
    const layer = layers[layerIndex];
    if (!layer) throw new Error(`edits[${index}] ${edit.op}: unknown layer ${pyRepr(name)}`);
    const poses = layer.poses;
    const target = edit.op === "set_pose" ? edit.pose : edit;
    const poseIndex = poses.findIndex((p) => p.angle === target.angle && p.frame === target.frame);

    if (edit.op === "set_pose") {
      const clonedPose = structuredClone(edit.pose);
      if (poseIndex === -1) poses.push(clonedPose);
      else poses[poseIndex] = clonedPose;
    } else if (poseIndex === -1) {
      throw new Error(
        `edits[${index}] ${edit.op}: no stored pose (${String(target.angle)}, ${pyNone(target.frame)}) ` +
          `in ${pyRepr(name)}; defaults are not implicit edit targets`,
      );
    } else if (edit.op === "delete_pose") {
      poses.splice(poseIndex, 1);
    } else {
      const pose = poses[poseIndex];
      if (pose) {
        pose.x = edit.x;
        pose.y = edit.y;
      }
    }
  });

  return PixelDefinitionSchema.parse(data);
}

/**
 * Partial port of `validated_art`. Python's version also builds a real `PixelArt` (from
 * `pixel_art.py`) and calls `art.validate_target(asset_layouts, frame_sequence, spec_dump())`,
 * which checks that authored poses actually match the configured canvases and contain visible
 * ink -- that requires the Canvas/PixelArt engine, which is ported in Phase 3
 * (`packages/pixel-core`); `packages/schema` has zero internal dependencies and can't reach it
 * yet (see `docs/typescript-rewrite.md`). This function therefore only reproduces the
 * schema-level half Python does before constructing `PixelArt`: `PixelDefinition.from_art` (drop
 * a stray `views` key, then validate) and `AssetSpec` validation, wrapped in the same
 * `DomainError` Python raises on any failure. **The engine-level `validate_target` check must be
 * wired in once `packages/pixel-core` exists** -- flagged here rather than silently omitted.
 */
export function validatedArt(
  data: Record<string, unknown>,
  options: { asset: unknown },
): PixelDefinition {
  try {
    const { views: _views, ...rest } = data;
    const definition = PixelDefinitionSchema.parse(rest);
    AssetSpecSchema.parse(options.asset);
    return definition;
  } catch (exc) {
    const message = exc instanceof Error ? exc.message : String(exc);
    throw new DomainError(`Invalid pixel-art definition: ${message}`);
  }
}
